polyfill/intersectionobserver-polyfill.js

```diff
@@ -0,0 +1,231 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+(function() {
+  // TODO: rootMargin are completely ignored for now
+
+  // Constructor
+  window.IntersectionObserver = function(callback, options) {
+    options = options || {};
+
+    if(!(callback instanceof Function)) {
+      throw('callback needs to be a function');
+    }
+
+    if(options.root && !(options.root instanceof HTMLElement)) {
+      throw('Target needs to be an HTMLelement');
+    }
+
+    this._callback = callback;
+    this._root = options.root || null;
+    this._rootMargin = options.rootMargin || [0, 0, 0, 0];
+    this._thresholds = options.threshold || 0;
+    this._init();
+  };
+
+  window.IntersectionObserver.prototype = {
+    //
+    // Public API
+    //
+    get root() {
+      return this._root || document;
+    },
+
+    get rootMargin() {
+      return '0';
+    },
+
+    get thresholds() {
+      // 0 means call callback on every change
+      // See note at http://rawgit.com/WICG/IntersectionObserver/master/index.html#intersection-observer-init
+      if(this._thresholds === 0) {
+        return 0;
+      }
+      if(this._thresholds instanceof Array) {
+        return this._thresholds;
+      }
+      return [this._thresholds];
+    },
+
+    observe: function(target) {
+      if(!(target instanceof HTMLElement)) {
+        throw('Target needs to be an HTMLelement');
+      }
+
+      var root = this.root;
+      var ancestor = target.parentNode;
+      while (ancestor != root) {
+        if (!ancestor) {
+          throw('Target must be descendant of root');
+        }
+        ancestor = ancestor.parentNode;
+      }
+
+      this._mutationObserver.observe(target, {
+        attributes: true,
+        childList: true,
+        characterData: true,
+        subtree: true
+      });
+      this._observationTargets.set(target, {});
+    },
+
+    unobserve: function(target) {
+      this._observationTargets.delete(target);
+    },
+
+    disconnect: function() {
+      this._observationTargets.clear();
+      this.root.removeEventListener('scroll', this._boundUpdate);
+      this._mutationObserver.disconnect();
+      this._descheduleCallback();
+    },
+
+    takeRecords: function() {
+      this._update();
+      this._descheduleCallback();
+      var copy = this._queue.slice();
+      this._queue = [];
+      return copy;
+    },
+
+    //
+    // Private API
+    //
+    _init: function() {
+      this._observationTargets = new Map();
+      this._boundUpdate = this._update.bind(this);
+      this.root.addEventListener('scroll', debounce(this._boundUpdate, 100));
+      this._mutationObserver = new MutationObserver(debounce(this._boundUpdate, 100));
```

Comment:
I think we want throttle here rather than debounce.  If the user scrolls smoothly and continuously down the page and the element we're watching transits the entire height of the viewport moving from out of view at the bottom to out of view at the top, the update fn will not be called at any point while the element is visible.

Also, there seems no reason to separately throttle the handlers for scroll and mutations, and finally the 100 is a magic number.  I'd suggest:

```javascript
this._boundUpdate = throttle(this._update.bind(this), POLL_INTERVAL);
this.root.addEventListener('scroll', this._boundUpdate);
this._mutationObserver = new MutationObserver(this._boundUpdate);
```

And then replace `debounce` with a minimal throttle implementation, such as:

```javascript
function throttle (fn, int) {
  var timer = null;
  return function () {
    if (timer) return;
    callback.apply(this, arguments);
    timer = setTimeout(function () {
      timer = null;
    }, int);
  };
}